api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+AVX2 has only gather instructions, no scatter instructions, whereas AVX512 has both. AVX2 is limited to 256-bit length vectors, whereas AVX512 has 512-bit support. Both support masking of individual memory accesses, using either a special mask register in AVX512 or another vector in AVX2.
+
+Examples of these instructions are (in DR’s IR):
+```
+vpgatherdd %rax(,%ymm11,4)[4byte] %ymm13 -> %ymm12 %ymm13
+```
+
+Above is an AVX2 gather instruction that reads 32-bit doublewords into the 256-bit `ymm12` vector from addresses generated by adding the base address in `rax` to the corresponding index elements in `ymm11`, conditionally based on the masks in `ymm13`. The mask is set to zero on completion, unless the instruction causes an exception.
+
+```
+vpscatterdd {%k1} %xmm10 -> %rax(,%xmm11,4)[4byte] %k1
+```
+
+Above is an AVX512 scatter instruction that writes 32-bit doublewords from the 128-bit `xmm10` vector to addresses generated by adding the base address in `rax` to the corresponding index element in `xmm11`, conditionally based on the mask reg `k1`. The mask register is set to zero on completion, unless the instruction causes an exception.
+
+
+# Problem Statement
+
+Scatter and gather instructions pose a challenge to DynamoRIO clients that observe memory addresses (like drcachesim, the DR client that collects memory address and control flow traces). They are complex to handle because:
+
+-  A single gather or scatter instruction loads from or stores to multiple addresses
+-  Accessed addresses may be non-contiguous
+-  Each access is conditional based on some mask
+
+DynamoRIO clients only see the scatter or gather instruction and need to do more work to extract all accessed addresses. This is unlike regular scalar loads or stores, where the accessed address is readily available. The goal of this work is to make it easier for DR clients to observe these addresses. We achieve this by expanding the scatter and gather instructions into a functionally equivalent sequence of scalar stores and loads. This way, DR clients will see regular store and load instructions which they can instrument as usual.
+
+Original issue: [DynamoRIO/dynamorio#2985](https://github.com/DynamoRIO/dynamorio/issues/2985).
+
+# Design
+
+This required the addition of new support in various DynamoRIO components, like drreg, drx, drmgr and core DR. Multiple contributors worked on designing and implementing the required changes.
+
+
+## Scatter/gather Instruction Expansion
+
+Owner: [Hendrik Greving](https://github.com/hgreving2304)
+
+As described above, we can simplify work for DR clients by replacing the scatter and gather instructions with a functionally equivalent sequence of scalar stores and loads. The expanded sequence is the unrolled version of the following loop:
```

Comment:
s/the ... instructions/each ... instruction/
(avoid ambiguity over replacing several scatters at once)

Reply:
Done